// ===================================================================== \\
// ===================================================================== \\
// ========================  _____ ========  _ 	© ====================== \\
// ======================== | ___ \ ======= | | ======================== \\
// ======================== | |_/ / __ _  __| |	======================== \\
// ======================== | ___ \/ _` |/ _` |	======================== \\
// ======================== | |_/ / (_| | (_| |	======================== \\
// ======================== \____/ \__,_|\__,_|	======================== \\
// ===================================================================== \\
// ===================================================================== \\
// Developed by Gabriel dos Prazeres - [Bad] © ========================= \\
// All rights reserved. Contact: +55 91 99322-5657 ===================== \\
// ===================================================================== \\
// ===================================================================== \\
// System: Bonus movement speed for vip ================================ \\
// ===================================================================== \\
// Compatible with: Emulator rAthena - 02/10/2021 ====================== \\
// ===================================================================== \\
// Description: ======================================================== \\
// Grants a movement speed bonus to vip player ========================= \\
// ===================================================================== \\
// Version 1.0 ========================================================= \\
// ===================================================================== \\
// Discord: Bad#5743 =================================================== \\
// ===================================================================== \\

From 38dee7da3ccbf006618a4d991da26e85c44bce67 Mon Sep 17 00:00:00 2001
From: Gabriel dos Prazeres <gabrieldosprazeres@hotmail.com.br>
Date: Mon, 4 Oct 2021 21:56:41 -0300
Subject: [PATCH] Bonus movement speed system for vip

---
 conf/battle/player.conf | 6 ++++++
 src/map/battle.cpp      | 1 +
 src/map/battle.hpp      | 1 +
 src/map/status.cpp      | 3 +++
 4 files changed, 11 insertions(+)

diff --git a/conf/battle/player.conf b/conf/battle/player.conf
index 456615ea1..a555b1645 100644
--- a/conf/battle/player.conf
+++ b/conf/battle/player.conf
@@ -233,6 +233,12 @@ vip_gemstone: 2
 // Will display rate information (EXP, Drop, and Death penalty message)? (Note 1)
 vip_disp_rate: yes
 
+// Movement speed increase.
+// Note1: 20 = 20% movement speed increase.
+// Note2: Set 0 for disable system
+// Default: 20
+vip_movement_speed_increase: 20
+
 // Revive dead player while warping? (Note 1)
 revive_onwarp: yes
 
diff --git a/src/map/battle.cpp b/src/map/battle.cpp
index b8024e111..cfadd67ed 100644
--- a/src/map/battle.cpp
+++ b/src/map/battle.cpp
@@ -8998,6 +8998,7 @@ static const struct _battle_data {
 	{ "vip_drop_increase",                  &battle_config.vip_drop_increase,              50,      0,      INT_MAX,        },
 	{ "vip_gemstone",                       &battle_config.vip_gemstone,                    2,      0,      2,              },
 	{ "vip_disp_rate",                      &battle_config.vip_disp_rate,                   1,      0,      1,              },
+	{ "vip_movement_speed_increase",        &battle_config.vip_movement_speed_increase,    20,      0,      INT_MAX,        },
 	{ "mon_trans_disable_in_gvg",           &battle_config.mon_trans_disable_in_gvg,        0,      0,      1,              },
 	{ "homunculus_S_growth_level",          &battle_config.hom_S_growth_level,             99,      0,      MAX_LEVEL,      },
 	{ "emblem_woe_change",                  &battle_config.emblem_woe_change,               0,      0,      1,              },
diff --git a/src/map/battle.hpp b/src/map/battle.hpp
index 1fd1e288c..210818771 100644
--- a/src/map/battle.hpp
+++ b/src/map/battle.hpp
@@ -556,6 +556,7 @@ struct Battle_Config
 	int vip_exp_penalty_base;
 	int vip_exp_penalty_job;
 	int vip_disp_rate;
+	int vip_movement_speed_increase;
 	int mon_trans_disable_in_gvg;
 	int emblem_woe_change;
 	int emblem_transparency_limit;
diff --git a/src/map/status.cpp b/src/map/status.cpp
index 5e572150e..7ac5d0e1d 100644
--- a/src/map/status.cpp
+++ b/src/map/status.cpp
@@ -7668,6 +7668,9 @@ static unsigned short status_calc_speed(struct block_list *bl, struct status_cha
 	TBL_PC* sd = BL_CAST(BL_PC, bl);
 	int speed_rate = 100;
 
+	if (pc_isvip(sd))
+		speed_rate -= battle_config.vip_movement_speed_increase;
+
 	if (sc == NULL || (sd && sd->state.permanent_speed))
 		return (unsigned short)cap_value(speed, MIN_WALK_SPEED, MAX_WALK_SPEED);
 
-- 
2.33.0.windows.2

